### PR TITLE
Add tracing for default probing for managed assemblies

### PR DESCRIPTION
**Pull request by elinor-fung**
- `KnownPathProbed` event for logging paths probed as described by https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/default-probing - possible sources of paths:
  - `TrustedPlatformAssemblies` : `TRUSTED_PLATFORM_ASSEMBLIES` property
  - `AppNativeImagePaths` : `APP_NI_PATHS` property
  - `AppPaths` : `APP_PATHS` property
  - `PlatformResourceRoots` : `PLATFORM_RESOURCE_ROOTS` property
  - `SatelliteSubdirectory` : satellite parent assembly's culture subdirectory 
- Added tests
- Moved some helper functions out of the BinderTraceTest class into a separate file/class

While maybe not so commonly used, I think the tracing for looking in the non-TPA probing path sources is helpful when they are used. For the TPA, 'probing' may be a bit of a weird term, since it is not really searching in some specific directory, just the full path from the property.

### CODE REVIEW

File: src/coreclr/tests/src/Loader/binding/tracing/AssemblyToLoad.csproj [hunk 1]
```diff
@@ -11,4 +11,7 @@
   <ItemGroup>
     <ProjectReference Include="AssemblyToLoadDependency.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <EmbeddedResource Include="Resource.fr-FR.resx" />
+  </ItemGroup>
```
Comment (lpereira) [inline src/coreclr/tests/src/Loader/binding/tracing/AssemblyToLoad.csproj]: Is this a related change?

Comment (elinor-fung) [reply src/coreclr/tests/src/Loader/binding/tracing/AssemblyToLoad.csproj]: It gives this assembly a satellite assembly (fr-FR culture) that the tests load.

File: src/coreclr/tests/src/Loader/binding/tracing/Helpers.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            return GetProbingFilePath(pathSource, assemblyName, isExe, null, null);
+        }
+
+        private static string GetProbingFilePath(ProbedPath.PathSource pathSource, string assemblyName, bool isExe, string culture, string baseAssemblyDirectory)
+        {
+            string baseDirectory = baseAssemblyDirectory ?? Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            if (!string.IsNullOrEmpty(culture))
+                baseDirectory = Path.Combine(baseDirectory, culture);
+
+            string extension = isExe ? "exe" : "dll";
```
Comment (lpereira) [inline src/coreclr/tests/src/Loader/binding/tracing/Helpers.cs]: This is set but not used. Refactoring remnant?

File: src/coreclr/tests/src/Loader/binding/tracing/Helpers.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        }
+
+        private static bool AssemblyNamesMatch(AssemblyName name1, AssemblyName name2)
+        {
+            if (name1 == null || name2 == null)
+                return name1 == null && name2 == null;
+
+            return name1.Name == name2.Name
+                && ((name1.Version == null && name2.Version == null) || name1.Version == name2.Version)
+                && ((string.IsNullOrEmpty(name1.CultureName) && string.IsNullOrEmpty(name1.CultureName)) || name1.CultureName == name2.CultureName);
```
Comment (lpereira) [inline src/coreclr/tests/src/Loader/binding/tracing/Helpers.cs]:
```suggestion
                && ((string.IsNullOrEmpty(name1.CultureName) && string.IsNullOrEmpty(name2.CultureName)) || name1.CultureName == name2.CultureName);
```

File: src/coreclr/tests/src/Loader/binding/tracing/Resource.fr-FR.resx [hunk 4]
```diff
@@ -0,0 +1,123 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<root>
+  <!--
+    Microsoft ResX Schema
+
+    Version 2.0
+
```
Comment (lpereira) [inline src/coreclr/tests/src/Loader/binding/tracing/Resource.fr-FR.resx]: `git commit -a`? I'd imagine you would use `fr_CA`, though. :)

File: src/coreclr/tests/src/Loader/binding/tracing/BinderTracingTest.csproj [hunk 5]
```diff
@@ -29,10 +32,15 @@
   </ItemGroup>
 
   <Target Name="MoveDependentAssembliesToSubdirectory" AfterTargets="CopyFilesToOutputDirectory">
+    <PropertyGroup>
+      <AssemblyToLoadResourceCulture>fr-FR</AssemblyToLoadResourceCulture>
+    </PropertyGroup>
```
Comment (lpereira) [inline src/coreclr/tests/src/Loader/binding/tracing/BinderTracingTest.csproj]: Related?

File: src/coreclr/tests/src/Loader/binding/tracing/Helpers.cs (hunk 2 above)
Comment (elinor-fung) [reply src/coreclr/tests/src/Loader/binding/tracing/Helpers.cs]: Doh. Yep, it should be used below.

File: src/coreclr/src/vm/ClrEtwAll.man [hunk 6]
```diff
@@ -7125,6 +7154,11 @@
                 <string id="RuntimePublisher.TieredCompilationSettingsFlags.NoneMapMessage" value="None" />
                 <string id="RuntimePublisher.TieredCompilationSettingsFlags.QuickJitMapMessage" value="QuickJit" />
                 <string id="RuntimePublisher.TieredCompilationSettingsFlags.QuickJitForLoopsMapMessage" value="QuickJitForLoops" />
+                <string id="RuntimePublisher.KnownPathSource.TrustedPlatformAssembliesMessage" value="TrustedPlatformAssemblies" />
```
Comment (vitek-karas) [inline src/coreclr/src/vm/ClrEtwAll.man]: There's a discussion that we should slowly start renaming TPA to something sensible: https://github.com/dotnet/runtime/pull/690#discussion_r355596277

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 7]
```diff
@@ -743,21 +743,22 @@ private static void OnAssemblyLoad(RuntimeAssembly assembly)
...
-            if (Internal.IO.File.InternalExists(assemblyPath))
-            {
-                return parentALC.LoadFromAssemblyPath(assemblyPath);
-            }
-            else if (Path.IsCaseSensitive)
+            bool exists = Internal.IO.File.InternalExists(assemblyPath);
+            if (!exists && Path.IsCaseSensitive)
             {
                 assemblyPath = Path.Combine(parentDirectory, assemblyName.CultureName!.ToLowerInvariant(), $"{assemblyName.Name}.dll");
+                exists = Internal.IO.File.InternalExists(assemblyPath);
```
Comment (vitek-karas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs]: Technically this is a second probe - so we should fire a second probe event.

File: src/coreclr/tests/src/Loader/binding/tracing/BinderTracingTest.DefaultProbing.cs [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+                Success = true,
+                ResultAssemblyName = asm.GetName(),
+                ResultAssemblyPath = asm.Location,
+                Cached = false,
+                ProbedPaths = new List<ProbedPath>()
+                {
+                    new ProbedPath()
+                    {
+                        FilePath = Helpers.GetProbingFilePath(ProbedPath.PathSource.AppNativeImagePaths, assemblyName.Name, isExe: false),
+                        Source = ProbedPath.PathSource.AppNativeImagePaths,
```
Comment (vitek-karas) [inline src/coreclr/tests/src/Loader/binding/tracing/BinderTracingTest.DefaultProbing.cs]: This is unfortunate side-effect of  `corerun`... normal apps executed through apphost or muxer would not have these. Maybe add a comment describing why this works the way it does.

File: src/coreclr/tests/src/Loader/binding/tracing/BinderTracingTest.DefaultProbing.cs [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+                        FilePath = Helpers.GetProbingFilePath(ProbedPath.PathSource.AppPaths, assemblyName.Name, SatelliteCulture.Name),
+                        Source = ProbedPath.PathSource.AppPaths,
+                        Result = S_OK
+                    }
+                }
+            };
+        }
+
+        [BinderTest]
+        public static BindOperation SatelliteAssembly_CultureSubdirectory()
```
Comment (vitek-karas) [inline src/coreclr/tests/src/Loader/binding/tracing/BinderTracingTest.DefaultProbing.cs]: While you're at it - can you add a test to validate that on Linux we perform case sensitive probe and then lower cased probe?

**Review by vitek-karas — APPROVED** [review]
The TPA naming in the events is my only concern - would be nice to make it consistent.

File: src/coreclr/src/vm/ClrEtwAll.man (hunk 6 above)
Comment (elinor-fung) [reply src/coreclr/src/vm/ClrEtwAll.man]: Good call. This one does at least have the advantage of not actually being part of the raw data / format that is traced, since it is just a string for the value map.
